### ISSUE TITLE
[META][D8] Add a reference module to core

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#1301]
**Description of the need**

I use node and user references all the time. It would be great if we could add these to core.

**Proposed solution (Updated 2022)**

Move the Entity Reference contrib module into core. It has been refactored to remove any external dependencies. Entity Reference is the most obvious candidate based on the following advantages:

1) It has the highest install count.
2) It has the most flexible API and number of extending modules.
3) It is a proven solution and includes an upgrade path for D7 Entity Reference users.

**Previous solution (Prior to 2022)**

_This was the original solution proposed in 2015. After the stalling of the new "Reference" project and the subsequent porting and updating of Entity Reference, this approach has failed to gain traction._

1) We should create a new module `reference` that is a rewrite of the Drupal 7 [entityreference](https://www.drupal.org/project/entityreference) module, in language that is simpler and easier to understand. It will more closely resemble `node_reference` and `user_reference` from the Drupal 7 [references](https://www.drupal.org/project/references) module. **There is now a new [Reference](https://github.com/backdrop-contrib/reference) module in contrib**
2) Once the module is ready, we should move this module into core. **When is the module "ready"?** When it has support for node references, and all [blockers in the contrib queue](https://github.com/backdrop-contrib/reference/issues/16) have been resolved. 

Core Blockers for adding the [reference module](https://github.com/backdrop-contrib/reference) into core
- [x] Add view function: https://github.com/backdrop/backdrop-issues/issues/1739
- [x] Define label and status entity keys: https://github.com/backdrop/backdrop-issues/issues/3103
- [x] Add options array for files: https://github.com/backdrop/backdrop-issues/issues/2299

We also have a single issue in the module queue documenting all the issues there that need to be resolved for core inclusion: [META: Blockers for getting into core](https://github.com/backdrop-contrib/reference/issues/36)

**Alternatives that have been considered**

* Put references into core: we decided not to put individual reference modules into core because in Drupal people seem to prefer having a single field type for everything. We would also prefer the consistency since we believe it will make a better user experience.

**Additional information**

Follow-up for merging in some bug fixes after this is merged: https://github.com/backdrop/backdrop-issues/issues/5665.

We have a follow-up issue for removing the term-reference field, and replacing it with a reference field, but it's likely that we'll get support for terms in first, long before we remove the previous term reference field: https://github.com/backdrop/backdrop-issues/issues/51. It's been recommended that we mark that field type as deprecated for 1 major cycle of backdrop core, and remove it in the next.

The core issue below is only needed for file entity support. If reference doesn't support file references at the time it goes into core, we don't need to handle this core issue yet, so it's not a blocker:
* Handle file stream wrappers : https://github.com/backdrop/backdrop-issues/issues/2300
-----

Advocate: @herbdool

**Comment by klonos** [comment]
:+1:

Besides field references, we should have a CKEditor addon (and/or a text format) so that people can add user/node references to a text field. Similar to how we add ```@user``` and ```#issue``` here on GitHub.

**Comment by herbdool** [comment]
I have a pull request in backdrop-contrib/references that fixes a few things for references, including the test. Other than that it seems to work fairly well.

**Comment by quicksketch** [comment]
I think it'd make most sense to make a single "References" module than it would to move User and Node Reference into core. Then we'd also need to convert Term Reference to References.

**Comment by mikemccaffrey** [comment]
Silly question, but why would build off of the reference module instead of the entityreference module?

Rather than having individual modules to reference each entity type, why not support a single entity reference field which allows optional targeting by type and bundle?

**Comment by mikemccaffrey** [comment]
Wow, nevermind about entityreference. We stand little chance of untangling that huge mess of ctools abstraction hell.

However, is it unreasonable for us to build a single reference field that works across all entity types? What are the challenges in the way of doing that?

**Comment by herbdool** [comment]
Not really a silly suggestion that entity reference would be better. Nate
said it might be a good thing to put into core. At least in some form. But
that will be a lot harder than using this module.

If and when references is in core I think it would be a good idea to add an
upgrade path from this one.

**Comment by jenlampton** [comment]
@mikemccaffrey the plan is to put references module into core because we like it's code (and dislike the entity reference code - sounds like you agree!)  However, we do want to include the *feature* entity reference provides. That will mean writing some new code for referenced module for a generic "reference a thing" field, then having a multi-step form for  "what kind of thing" and "what kind of kind of thing" etc etc.

**Comment by klonos** [comment]
> ...the plan is to put references module into core because we like it's code (and dislike the entity reference code...

I don't think I disagree here, but from an end-user point of view - not codebase-wise. I'll have to test them both again to be sure I have a better idea of their features and UX. Just wanted to point the fact that [Entity reference](https://www.drupal.org/project/entityreference) has been merged into D8 core:

![drupal-entity_reference-merged_into_d8](https://cloud.githubusercontent.com/assets/2423362/11011857/c0fd35c8-8540-11e5-9695-b2a0634ba0a0.png)

I realize though that we are trying to be in parity with D8 feature-wise and not necessarily architecture-wise, plus this is a goal and not a promise, so best effort. So, as I said, just pointing the fact out.

**Comment by mikemccaffrey** [comment]
Hey folks, I'd just like to reiterate that I'm very interested in working on this issue and to create a reference module that combines the best of all the different existing implementations, and perhaps leverages the relative-simplicity of backdrop to streamline the code and interface.

Who else is interested in helping out with this effort?

**Comment by gifad** [comment]
Two thoughts on references : 
1. Reference field is not very usefull without a useable "items" browser, where you can drag "items" from - so I'd design the browser first, the field second.
2. EntityReference benefit is mitigated if a given field instance can reference only a single entity type - else it is forced to a high level of abstraction, plugins, etc...
My point of vue is to node everything, so the browser can be a simple view, and we already have a node reference...

**Comment by serundeputy** [comment]
@mikemccaffrey one idea is if you just make a fork/port of node reference or whatever starting point you want and start to architect it.

once you see the challenges and identify tasks; you can make issues in your repo and assign/appeal for help to willing developers.

I for one am looking for something to sink my teethe into for the `1.4` cycle, but I'm not sure if that will be redirect, focus on drush or both, but if there are sort of bite size issues in your repo we can all chip in. I say you be the architect and see what you can dole out.

**Comment by mikemccaffrey** [comment]
> @gifad: 1. Reference field is not very usefull without a useable "items" browser, where you can drag "items" from - so I'd design the browser first, the field second.

Well, the data architecture should be pretty simple, so there is no reason not to tackle that first. Basically, we just need to store the entity type and entity id being referenced. The UX needed to help the users find the content they are looking for are the tricky part. For the time being I'd like to start with the autocomplete on the entity title field, but we can certainly also try to develop a more visual browser.

> @gifad: 2. EntityReference benefit is mitigated if a given field instance can reference only a single entity type - else it is forced to a high level of abstraction, plugins, etc...

Actually, we already have a good deal of abstraction in the form of the Entity API. The information coming from hook_entity_info gives us all the entity types, their bundles, their primary key, and base table. And we can perhaps add some additional keys to those return arrays so entities in backdrop can also indicate which fields and values control things like their published status.

> @serundeputy: one idea is if you just make a fork/port of node reference or whatever starting point you want and start to architect it.

Well none of the modules seem like an especially good starting point, so I think I am going to start with a new project and copy the best parts out of each of the implementations. I'll probably start it in my own github account, move it to contrib when it is ready for testing, and then we can move it into core for 1.4.0.

**Comment by mikemccaffrey** [comment]
Okay, here is the very very beginning of the new module, if anyone is interested:
https://github.com/mikemccaffrey/reference

Once I get the basic functionality working, I'll move it to contrib so others can help out.

**Comment by klonos** [comment]
Exciting! Let me know when it is ready for a general spin.

**Comment by MrHaroldA** [comment]
> Who else is interested in helping out with this effort?

Where do I sign up? ;)

**Comment by jenlampton** [comment]
👍

**Comment by jenlampton** [comment]
Pushing to 1.5. But let's get it done in contrib if ready early.

**Comment by quicksketch** [comment]
So in our weekly meeting today (https://www.youtube.com/watch?v=GsORuW80I3U) we officially pushed this to 1.5 after the status update. In the mean time, @mikemccaffrey is going to move the References project into the backdrop-contrib group on GitHub, where we can collaborate together on it as a contrib project, then move it into 1.5.0 (the release after this one), which will be September 15, 2016.

**Comment by MrHaroldA** [comment]
I've created a couple (temporary) repositories for what I've cooked for my Domains-like module; including a simple reference implementation. Is this of any use?

https://github.com/MrHaroldA/reference
https://github.com/MrHaroldA/domain

**Comment by herbdool** [comment]
@MrHaroldA might be some use to collaborate with @mikemccaffrey to get a contrib version working. Looks like it's not getting into 1.4 but will live as a contrib module until then. Before it gets into core it'll need some tests also.

**Comment by mikemccaffrey** [comment]
I've moved the module that I have been working on into backdrop-contrib:
https://github.com/backdrop-contrib/reference

Once we get it working there, we can move it into core.

**Comment by biolithic** [comment]
Thanks Mike!

**Comment by mikemccaffrey** [comment]
Hey folks,

I've been trying to figure out a better way to allow the reference field to reference all different sorts of entity types and bundles as needed. I think I've hit upon a workable way, where on the field instance settings form, you check off all the entity type / bundle combinations that should be referenced. Then in the selection widget on the edit forms, the user would select the type of content to be referenced, and then use the autocomplete to find the right item.

Here is a screnshot of a field instance settings form, showing the widget in the default value section:
![reference_entity_bundles](https://cloud.githubusercontent.com/assets/1134098/17417082/623ec5b6-5a46-11e6-94b9-0a05ff34bc57.png)

Let me know what you think of this approach!

**Comment by quicksketch** [comment]
Thanks for sharing the screenshot! My first thought on this is that the list of Entity types + Bundles could get very long very quickly. With each content type adding 2 items (comment and node), plus other types of entities that may be added in the future (files, etc).

You mentioned during weekly meetings that we also have a problem with the auto-complete only working for a single type of entity. Is that the case when creating content using this field as well as when setting the default value?

Overall, I'm not sure we need the ability to reference multiple types of content *within the same field*. I've never run up against such a situation. Even if I had the choice, I feel I might be inclined to make two separate fields anyway just to help the situation of auto-completing in a sane way. I might end up with user accounts and nodes that have the same title for example, I'd want to be able to select each separately rather than them all being in the same list. Displaying threaded together different types of entities also seems like it would present problems, how would you select the display modes for different types of entities (maybe you've already thought of that though). What do you think? Are we overshooting here?

**Comment by herbdool** [comment]
This looks like a 2.0 feature. For 1.0 can we aim for something that is robust but with fewer features? Being able to have autocomplete references to any kind of content or files sounds awesome but I do find it hard to know of a use case. Doesn't mean it doesn't exist, but likely will be in the 20% or less of people using reference.

**Comment by klonos** [comment]
Well, there is a very simple case where being able to lookup in multiple tables would be useful...

You see, with "simple" references we can for example reference a user from a node, but if we need to do the opposite, like reference nodes of various content types, then we'd need to be able to lookup in various tables. Right?

Consider the backdropcms.org case: creating a user or Organization profile and wanting to link those to any contribution they've made. We would need a reference field that would allow multiple values and that would be able to do a lookup in module nodes, theme nodes, layout nodes and perhaps also events, blog posts, comments, documentation nodes, ...perhaps more.

That would be a valid use case and IMO not a 20% one. Of course, we can get a "simple" version of reference in for 1.5.0 and then iterate on improvements, it's just that it'd be best if we build it in such a way that adding those improvements does not require API breaking (and thus improvements can be added in the 1.x cycle).

My 2c

**Comment by klonos** [comment]
> My first thought on this is that the list of Entity types + Bundles could get very long very quickly.

...let me ponder on that. Perhaps I can either find a common pattern of something that we (Drupal/Backdrop) do in such cases or come up with a new idea.

**Comment by klonos** [comment]
I googled around for a solution to this, but was unable to find something ready-made that we could use 😞 

My idea was to use a "mutated" version of either a fieldset or an accordion that shows/hides parts of it content (based on attributes). Mockups upcoming...

**Comment by herbdool** [comment]
In your example @klonos wouldn't the linkage be between the user account that's recorded with the node or comment? Even with a profile a Views relationship can be made to get the user related to the profile and get all the content. Maybe I don't understand the example completely.

At any rate, from what I understand, being able to use a View as a reference was already possible in the older Drupal modules so I figure that would be a good to have here as well. In that case, any/all content types could be referenced. The key barrier, though, is crossing between entities. E.g between files/users/content type/ etc.

I'd prefer that we replicate the functionality that's already present in Entity reference and/or References before trying to add on to it.

**Comment by klonos** [comment]
...so the user loads the form and they see no bundles configured to be referenced by default:

![backdrop-issue1301-mutated-nested-fieldset-1-unconfigured](https://cloud.githubusercontent.com/assets/2423362/17446711/da7eea86-5b8d-11e6-83e6-b3bfe89375d0.png)

...they click the "add bundles" fieldset end they are presented with a (long) list of available bundles to select from:

![backdrop-issue1301-mutated-nested-fieldset-2-long-list-of-available-entity-bundles](https://cloud.githubusercontent.com/assets/2423362/17446765/28880ed8-5b8e-11e6-8fee-1bd5d687acfd.png)

...lets say they randomly select a few of them:

![backdrop-issue1301-mutated-nested-fieldset-3-randomly-select-some-entity-bundles](https://cloud.githubusercontent.com/assets/2423362/17446811/705364a6-5b8e-11e6-8f43-5eadf628bedf.png)

...after saving the form, the selected entity bundle checkboxes bubble up and are rendered outside the collapsed fieldset that holds the rest of the available (but not selected) entity bundle checkboxes:

![backdrop-issue1301-mutated-nested-fieldset-4-selected-entity-bundles-buble-up-always-visible](https://cloud.githubusercontent.com/assets/2423362/17446889/d4a276ae-5b8e-11e6-97be-77d29f700612.png)

This way, when the user scrolls through that form, they are still able to see an overview of the selected entity bundles (no extra clicks = better UX), but at the same time, the long list of the rest of the available bundles is not in their way (less scrolling = better UX).

...lets say they need to remove a bundle and add another:

![backdrop-issue1301-mutated-nested-fieldset-5-add-remove-entity-bundles](https://cloud.githubusercontent.com/assets/2423362/17446976/567ee608-5b8f-11e6-89a0-edfccfa69ab6.png)

...after saving the form, the new bundle(s) bubble up outside the collapsed fieldset and the disabled ones are hidden in it:

![backdrop-issue1301-mutated-nested-fieldset-6-entity-bundles-updated](https://cloud.githubusercontent.com/assets/2423362/17447041/b1e0926c-5b8f-11e6-936e-650cd59eca6e.png)

![backdrop-issue1301-mutated-nested-fieldset-7-entity-bundles-updated-fieldset-expanded](https://cloud.githubusercontent.com/assets/2423362/17447052/bbd0a168-5b8f-11e6-99a1-fa13653af404.png)

I believe that such an implementation would address your concern about a long list of bundles @quicksketch. @wesruv thoughts?? Others?

**Comment by jenlampton** [comment]
Before we get too excited mocking up UI's, I'd like to remind everyone that we should avoid - **at all costs** - using either the word "entity" or the word "bundle" in any user interface.  Below is an example most human beings would be able to understand:

"What would you like to reference?"
- Content
- User account
- Category

"Limit X by type"
- Post
- Page

> Overall, I'm not sure we need the ability to reference multiple types of content within the same field. I've never run up against such a situation.

I also agree that we should not seek to over-engineer the initial module for core. We have 2 solutions for this in Drupal, neither of which include this feature. Though it might be nice to have it is most certainly not a requirement and if it presents any sort of challenge at all we should drop the feature. It's not very important, if it were, it would already exist :)

**Comment by klonos** [comment]
@herbdool I was thinking this use case:

- Project nodes in backdropcms.org are actually 3 separate node types (modules, themes, layouts)
- Project nodes are automatically created by a non-actual user (the "drop").
- A registered user would want to reference from their profile which projects they maintain or have contributed code or documentation to.

...in this scenario, which is an actual use case, the user is not creating the nodes, so that info cannot be pulled that way. We'd want to provide a "projects contributed to" multi-value reference field that would allow the user to autocomplete search across module, theme and layout nodes.

Although you might argue that this is likely a design flaw (project nodes could have been set to be a single content type for example, and the distinction between project type could have been a taxonomy tag), still this is a real-life use case where you can't ask the site owner to redesign their content types and migrate their content in order to be able to use references. We'd need to accommodate that.

**Comment by klonos** [comment]
>...we should avoid - at all costs - using either the word "entity" or the word "bundle" in any user interface.

@jenlampton 👍 of course. The mockups were drafted in a few minutes as prof of concept for dealing with long list of "things to be referenced" 😄

**Comment by klonos** [comment]
...the goals with the mockups was to keep the list out of the users' way (less scrolling), but at the same time provide an overview of already selected "things to be referenced" without requiring additional clicks.

**Comment by jenlampton** [comment]
> Although you might argue that this is likely a design flaw (project nodes could have been set to be a single content type for example, and the distinction between project type could have been a taxonomy tag), still this is a real-life use case where you can't ask the site owner to redesign their content types and migrate their content in order to be able to use references. We'd need to accommodate that.

This is a non-issue. All existing reference fields already allow for this relationship and cutting the new feature mike has described does not affect our ability to do it.

**Comment by klonos** [comment]
@jenlampton perhaps I did not understand correctly what is the feature that @mikemccaffrey was talking about that you and @quicksketch believe should be left out. I thought that that feature was the ability to be able to search on entity autocomplete by targeting multiple entities...

> Overall, I'm not sure we need the ability to reference multiple types of content within the same field.

If that was not implemented, then how would one (in the example of backdropcms.org project nodes) be able to search through a *single*, multi-value autocomplete field against module nodes, theme nodes and layout nodes?

Perhaps haven't been getting much sleep lately has started to effect my brain 😜

**Comment by klonos** [comment]
...perhaps that is what @herbdool was referring to when he said

>...being able to use a View as a reference was already possible in the older Drupal modules so I figure that would be a good to have here as well. In that case, any/all content types could be referenced.

That way (again in the specific use case of backdropcms.org that I have in mind), one could create a view of content of all type, then filter by content type = module AND theme AND layout, and then use that view as the reference auto-complete. Maybe? Right?

**Comment by klonos** [comment]
Are we talking about the same thing? 😄 💤

**Comment by herbdool** [comment]
I think that's what I'm saying. Existing modules allow you to use multiple node bundles, but not across entity types (can't combine users plus files plus users plus category).

**Comment by klonos** [comment]
That makes better sense then 😄 

To be fair, although at the moment I cannot think of any case where targeting multiple entities would be of use, it still does not seem far-fetched.

**Comment by jenlampton** [comment]
> To be fair, although at the moment I cannot think of any case where targeting multiple entities would be of use, it still does not seem far-fetched.

I agree! However, if it presents a significant challenge it should be dropped. it's not as important as single-entity references.

**Comment by jenlampton** [comment]
CRAZY IDEA: Since we are running out of time for writing a new module for Backdrop 1.5, I want to propose instead that we add the (already ported) references module to core. This is something we can actually do in the 3 weeks we have remaining before release of 1.5, and meets all our requirements for the feature we need to add.

After References is included, we can work on adding more entity-reference like features, and an upgrade path for node/user/term reference in the future, if needed.

**Comment by herbdool** [comment]
I'm all for "minimal viable product" and incremental improvements and seems to fit well with Backdrop philosophy. References works, it can reference two types of entities which covers a lot of use cases and it is possible to provide upgrade paths as it gains new functionality to reference more types of entities and perhaps different kinds of widgets.

On the con side, it might make it harder to port modules from Drupal that rely on entity reference (though this is true of the Reference module in progress as well.

**Comment by docwilmot** [comment]
If it helps, I stripped the CTools from Entity Reference and got it to work normally on minimal testing. I used direct function calls to replace the plugins (should be hooks?), and cant figure out a replacement for `entity_access` (so just returned TRUE), but those were the only two major issues. Also the tests and examples werent checked.
Have a look https://github.com/docwilmot/entityreference. 
Minimal testing mind you.

**Comment by jenlampton** [comment]
@docwilmot the problem is that we don't like the code style in entity reference, so we wouldn't want to put in a direct port without a major rewrite. The rewrite is what we don't have time for, so we're looking for something that would be ready to go in with minimal work.

**Comment by docwilmot** [comment]
@jenlampton my thinking is that if we're going to put in a minimally viable references module this one is the better option. If we put in References, 
- it won't do what we want, i.e. give the ability to reference any entity with a single field type
- we'll eventually need to remove it again, including changing the API and UI, and
- we'll need upgrade paths for user's existing node and user references

If we now minimally modify Entity Reference to taste and put it in, we can incrementally/drastically improve/change the codebase without anyone in the front end being the wiser.

**Comment by herbdool** [comment]
@docwilmot has a point. The end user nor the site builder probably doesn't care about the code style.

Another advantage of Entity Reference is that modules like: https://www.drupal.org/project/entityreference_prepopulate, https://www.drupal.org/project/entityreference_view_widget, http://drupal.org/project/entityreference_dynamicselect_widget are easier to port. (Though it looks like https://www.drupal.org/project/multiple_selects will work with references too).

**Comment by serundeputy** [comment]
@docwilmot @herbdool those are fair and valid points, but it does not change the fact that using the Entity Reference path will not be able to be completed in the `1.5.0` cycle.

The choice here as I see it is:

1. Choose `Reference` brand new module and push to `1.6.0`
2. Choose `Entity Reference` and clean up code style and push to `1.6.0`
3. Choose `References` which has upgrade path and working code; Leave in  `1.5.0` cycle and iterate to bring in the other aspects of `Entity Reference` and `Reference`

I think option `3` adds immediate value to our users, but will require more effort by devs to enhance the functionality without breaking UIs and APIs.

**Comment by jenlampton** [comment]
I tried to look at entity reference again yesterday to see how bad the code
was, see if maybe we couldn't make a push on that for 1.5, but it
immediately gave me a headache. Someone else want to see if there are any
easy abstraction reductions we can do without breaking it?

Jen

- -
Please excuse brevity and typos. Message typed with my thumbs from an Open
Source OS.

**Comment by jromine** [comment]
@jenlampton yes, this is why I initially ported references instead.

**Comment by herbdool** [comment]
Seems that @docwilmot has a working module, though?

**Comment by jenlampton** [comment]
@herbdool there's a big difference between "working" and "core quality".  We can't just add a working module to core because it's closer than all the alternatives. Most modules get a rewrite when they go into core, and upgrade paths are tested & included.

I'm hearing concerns about how to get off of node_reference, user_reference, and term_reference if we add `references` now. Please keep in mind that we don't get away from that problem if we go with `entity_reference` or `reference` instead. We still need the upgrade path, it's just a matter of whether we need it at the time we put something in core, or if we can do it later on when we improve the something that's in core.

* With `references` the rewrite will be minimal, and the upgrade paths are non-existent since the data is in the same exact structure it was in Drupal 7. **We have time for this.**

* With `entity_reference` (even if we decided not to rewrite it at all) would still need an upgrade path for the term_reference module that's in core, as well as from `references` in contrib. **It's most likely we don't have time for this either** But then we'd also need to live with the code that is in entity reference as it is. Not ideal.

* With `reference`, the module is not done, and it needs an upgrade path from references, term_reference, and entity_reference. **We do not have time for this** This is the best possible approach, which is why we tried it. 

It sounds like there's some concern about my crazy idea (and understandably!).  If we are going to push a reference module to 1.6 instead of 1.5 I'd rather stick to @mikemccaffrey's approach and move forward with `reference`. (I also like it's namespace better :)

**Comment by quicksketch** [comment]
Hi all, no updates here in the issue, but we discussed this in person during the last weekly meeting: https://www.youtube.com/watch?v=MAMIaVeyr1g

A little banter in there were we weighed both sides. But ultimately we punted this to the PMC for decision making. By unanimous decision we're moving this to 1.6.0 rather than trying to rush something through.

I think Mike's new Reference module is the best bet we have, but we have a lot more work to do here (upgrade paths especially). Let's continue working on this for the next release. The Roadmap has been updated to reflect this: https://backdropcms.org/roadmap#1.6.0

**Comment by jenlampton** [comment]
I'd like to commit to working on an upgrade path from term_reference starting around Dec 15th. I've been dealing with issues like [this one](https://www.drupal.org/node/2557307) in Drupal, and the more I think about it, I can't see any good reason to leave two ways to do the same thing in core. It's only going to create confusion for users, and extra work for developers to have to support two different APIs to provide one feature in contrib. We need to kill term_reference.

**Comment by jenlampton** [comment]
I know @mikemccaffrey has been working hard on getting tests ready for References, but I haven't gotten around to upgrade paths yet. Tonight is feature freeze, so I don't think references is going to make it for 1.6.  

We do have a solid [contrib module](https://github.com/backdrop-contrib/reference) that can be used today. We can work there on upgrade paths, and continue making progress in contrib until we can move references into core.

I'm going to move the milestone from 1.6 to 1.7. If the contrib project is ready by the next release I'd like to consider it again for inclusion.

**Comment by klonos** [comment]
>Tonight is feature freeze, so I don't think references is going to make it for 1.6.

As I also suggested during our past dev meeting, lets please favor a stress-free 2017 for contributors over getting things done. I am totally fine with taking our time and getting things done the right way the same way we pushed WYSIWYG to next releases until we felt confident about it.

@mikemccaffrey 👍 and please keep up the good work and do not let deadlines bring you down. I am sure you'll eventually get there. I just wish I could help.

**Comment by jenlampton** [comment]
Since references now has a stable contrib release, we can work out issues there and re-schedule this for core inclusion when we feel it's ready. Removing the core release milestone.

**Comment by BWPanda** [comment]
Just closed https://github.com/backdrop/backdrop-issues/issues/51 as a duplicate of this one.

**Comment by stpaultim** [comment]
We've had several dev meetings over the last several weeks on the topic of development priorities for the 1.21.0 release. This issue has gotten a lot of attention. @hosef has indicated that quite a bit of progress has been made along with a scaling back of expectations regarding what must be accomplished before this module gets into core. 

There is some mixed feelings about whether or not it's realistic to get this into the 1.21 release of core.

It is my understanding, that adding this to core simple means:

1) Testing and fixing the current contrib module
2) Dropping this into the core modules directory

Folks can help us get this into core by doing the following:

1) Installing and using the stable release of this module and reporting bugs. https://backdropcms.org/project/reference
2) Looking through the issue queue for this module and helping with any issues found there. https://github.com/backdrop-contrib/reference/issues

@hosef - can you say anything more about where help is most needed?

I'm inclined to schedule a sprint day focused on this module - if there is interest and anyone thinks it might be helpful.

Getting this into the 1.21.0 release would be a pretty big victory, in my opinion.

**Comment by herbdool** [comment]
I'm not convinced it's ready.

In this issue summary it lists core blockers that prevent this from going into core. Plus there are blockers in the contrib issue queue.

I've brought up issues that need to be resolved before I'm convinced it's ready too.

**Comment by philsward** [comment]
As much as I would love to see one module replace all others and included in core, I feel like it needs a lot more scrutiny before including it. It would be a great addition for new sites, but I think it needs to be ready for the migration of existing BD and potentially D7 sites to be considered for core.

When I played around with references a month ago or so to replace term_reference, it was obvious it isn't ready to replace existing methods (I don't remember the specifics off hand). I ended up using term_reference instead. Surprisingly, references is a lot farther along than I figured, but it still isn't quite ready.

Point is, this should, as contrib, do everything the other modules do, then be tested for migration from the other modules to this one before being included in core. 

This isn't a core module to take lightly.

I would be OK with it not being prime-time for a seamless D7 to BD migration (that could come later) but at least a solid migration from the other iterations on BD to the new BD iteration should be solid before including it in core.

That's my take on it. I'm good either way and will do what I can to help test to move things along.

**Comment by klonos** [comment]
Back in July this year, in https://github.com/backdrop/backdrop-issues/issues/1739#issuecomment-877487307 @jenlampton mentioned that that was the only blocker.

**Comment by klonos** [comment]
> ... we may also need a way to block the install of references and entity references from the project installer ...

When we merge contrib modules into core, we take care of this as part of the task. We have the `backdrop_merged_modules()` function for that + we usually update the respective contrib module with something like `dependencies[] = system (<1.21)` which makes it "incompatible" with newer versions of core.

**Comment by herbdool** [comment]
> ... we may also need a way to block the install of references and entity references from the project installer ...

No way should we block those other modules from being installed! Reference is incomplete and doesn't have an ecosystem that works with it.

Personally I still don't understand why we don't just put entity reference into core instead of trying to recreate it all. There's been some vague references to the code base, but I don't understand the argument. Entity Reference works in Backdrop, has other contrib modules that work with it or can be easily ported. And any issues people have with the codebase can be fixed. If someone will just point them out specifically.

Reference is not ready. And there are multiple blockers as I pointed out in the module's issue queue.

**Comment by olafgrabienski** [comment]
> No way should we block those other modules from being installed! Reference is incomplete and doesn't have an ecosystem that works with it.

I agree, and if I recall correct, the question has been discussed somewhere with the result to not block other modules without having reached feature parity and a working upgrade path. Does someone remember the discussion and can provide a link to it?

**Comment by philsward** [comment]
@herbdool @olafgrabienski I should have clarified... I suggest to block those only after reference is in core and the appropriate action for site builders is to migrate the contrib to the new core module because the contribs are considered obsolete. Not before it's in core.

@herbdool, it was mentioned a while back that entity_reference is a spaghetti of code that is difficult to incorporate, especially with its heavy reliance on ctools. I don't personally know or understand the details of this though. I think @jenlampton can offer some insight on this?

**Comment by herbdool** [comment]
@philsward entity reference is already ported and doesn't rely on ctools any longer so that part isn't an issue. (It'ss not the only part of core that used to rely on ctools). I help maintain it and don't think the code is any worse than some parts of core. It's definitely not spaghetti code.

**Comment by jenlampton** [comment]
> Personally I still don't understand why we don't just put entity reference into core instead of trying to recreate it all.

Here's the reasoning from the Original Post:

>  Put entity_reference into core: we decided that the code for entity_reference module in Drupal 7 was too obfuscated for Backdrop. It depends on the Entity API most of which is not strictly necessary, as long as you use the core APIs that are provided. A rewrite would allow us to create a module with identical features that is simpler and easier for everyone to contribute to.

The current problem is that entity_referece depends on entity_plus, but everything is possible with core alone. The unnecessary layers of abstraction in entity_plus make the code harder for people to work with, reducing the number of people who will be able to contribute to it. 

I haven't looked at the code for a while, so I don't have any specific examples, but I do know that entity_reference is still missing features that were present in both references and term_reference, and if the code had been approachable and easy to work with, those features probably would have been included by now.

**Comment by herbdool** [comment]
The only reason *entityreference* depends on entity_plus is because core does not yet support `entity_view()`, which is something that *reference* is also blocked on (https://github.com/backdrop/backdrop-issues/issues/1739). There is one other, *optional* call to entity_plus but that wouldn't have to go into core. So removing dependency on entity_plus is no more troublesome for entityreference.

If the focus had been on improving entityreference UI instead of creating something from scratch, we'd probably be much further along by now.

**Comment by docwilmot** [comment]
I also feel we should move Entity Reference to core. No point struggling (and apparently failing, after all these years) to build a new module from scratch, when this one has been proving itself on D7 and Backdrop for years. We dont `entity_view` to be in core first either.

**Comment by jenlampton** [comment]
I thought it was the metadata wrapper madness that was why it needed entity_plus, and that definitely doesn't belong in core. I wasn't aware of the `entity_view()` issue, but that one seems like it could be an easy fix. 

If entity reference doesn't use metadata wrappers (I might be thinking of something else) and everyone else is in favor of putting entity reference in core instead of reference, we should probably start a new issue to manage the todos for what that process would look like. I don't think anyone has seriously considered it since we've always been working toward reference. 

Just noting here that we would be setting aside our second [principle](https://backdropcms.org/philosophy#principles) in favor of the third. (which might be okay?)

**Comment by hosef** [comment]
Ok, I spent some time reading through this issue and several other old issues that were related.

Some things I would like to point out before moving on:
- This issue is over 6 years old
- The discussion about references vs entityreference vs a new module happened before any of the modules in question had been ported
- The issue was stalled for almost 4 years before I started working on it.
- I started working on it because I thought everyone had agreed about which direction to go in

As I understand the history of the issue, there were 3 main problems with using the entityreference module:
- The people involved did not want to port the entityreference module
- The people involved did not want to show the word ‘entity’ in the user interface because it does not mean anything to non technical users of Backdrop
- The people involved did not want to conflict with the name space of an existing reference module

If people now think that the entityreference module is stable enough to be the default reference solution, we could just rename the functions, update the UI, and call it a day.

As far as entity metadata wrapper goes, I think that is way more simple than what we currently do to look up field values from entities.

**Comment by hosef** [comment]
As far as entity metadata wrapper goes, I think that is way more simple than what we currently do to look up field values from entities.

**Comment by philsward** [comment]
One implementation to rule them all is what I'm in favor of.

How we get there or why is less important to me than making sure whatever does get included, is rock solid.

**Comment by indigoxela** [comment]
One thing to consider: what do people actually use.

- Entity Reference:  214 Active installs
- Reference:  9 Active installs
- References:  301 Active installs

It seems like the "Reference" module has the fewest installs.

Some personal feedback: one key feature for me is, that I can use a view as base for referenced nodes/users/terms.

**Comment by philsward** [comment]
Does it make more sense to go off BD installs or Drupal installs?

**Comment by olafgrabienski** [comment]
> Does it make more sense to go off BD installs or Drupal installs?

In my opinion, both are interesting. Here is a current snapshot from the [Drupal statistics](https://www.drupal.org/project/usage), both for D7:

- Entity Reference: 192,011 installs
- References: 68,783 installs

Both modules are quite popular. Interesting that Entity Reference is more popular in Drupal but less popular in Backdrop.

**Comment by laryn** [comment]
Note: Judging popularity is a bit of guesswork depending on when a port was done or an official release made (so that it showed up in the installer)

**Comment by indigoxela** [comment]
> Interesting that Entity Reference is more popular in Drupal but less popular in Backdrop

Keep in mind that both modules have not been released at the same time. References has been ported 2016 and Entity Reference in 2019. So the latter is available for a way shorter time.

Ah, @laryn just beat me to it :stuck_out_tongue_winking_eye:

**Comment by BWPanda** [comment]
Here're my thoughts as a core committer:

`reference` has/is being written for Backdrop from the ground-up. It doesn't have any old, legacy code, no baggage, etc. It's clean, simple, and easy to maintain going forward. 'entity_reference` on the other hand is a Drupal port and, as such, is complex, has extra baggage, dependencies, etc. and raises the barrier-to-entry for new/existing maintainers.

For people coming to Backdrop from Drupal, they have options in contrib that match what they're already using in Drupal. So they're taken care of. Therefore, in my mind, an option in *core* is specifically for people building a new Backdrop site from scratch. For them, I'd much prefer to see the clean, simple solution that is `reference`, even if that means it takes longer to finish and include (again, we have working contrib modules for the time-being, and down the track I'm sure there'll be an 'upgrade' process from contrib's `entity_reference` to core's `reference`).

**Comment by herbdool** [comment]
> It doesn't have any old, legacy code...

Isn't almost all of Backdrop "legacy code"?

There's a big benefit to legacy code where a lot of exceptions and bugs have already been encountered and fixed.

**Comment by MrHaroldA** [comment]
> There's a big benefit to legacy code where a lot of exceptions and bugs have already been encountered and fixed.

Only if that code is completely unchanged, and used on the same framework it was written on. Since it's a port, I'd say that it is unstable until proven otherwise.

**Comment by klonos** [comment]
For the record, I agree with what @BWPanda said. We should aim for simplicity and allowing more than a handful of people to be able to contribute to whatever ends up being in core. People with existing D7 sites can use the 2 existing equivalent contrib modules when the move to Backdrop, and when we eventually have an upgrade/migration path in place, they can move to the core-provided solution if they choose so.

**Comment by herbdool** [comment]
@MrHaroldA not exactly starting from zero here. There are over 200 installs. But we're not just talking about entityreference here, but also comparing it to reference.

I find it curious that people are making a big deal about entityreference when core has a lot of examples of equally "complicated" modules. I'm not exactly sure what complexity people are pointing to in entityreference (would be good to get some actual examples from objectors). Layout has a complicated system of classes. Views uses a system of plugins for "access", "argument", "display" that extend the views_plugin class that is quite similar to how entityreference has plugins for "behavior" and "selection". Surely no one is arguing we "simplify" Views as well.

I'm not even sure what simplicity would look like in these examples. Does it mean that only procedural programming should be used? (I'm skipping over Entity Meta Wrapper since I've already pointed out entityreference doesn't actually rely on it.)

This issue seems to be the big exception to the codebase in Backdrop. We've ported and simplified where possible, and also maintaining most of the backwards compatibility.

**Comment by philsward** [comment]
As of Nov 7, 2021

Reported [Drupal 7](https://www.drupal.org/project/usage/drupal) Installs: **532,890**
Reported [Entity Reference](https://www.drupal.org/project/usage/entityreference) Installs: **191,555**
Reported [References](https://www.drupal.org/project/usage/references) Installs: **68,129**

Combined installations of a module that deals with referencing: **259,684**

Nearly half of all D7 sites use some sort of module for referencing entities. This doesn't include those who are only using Term Reference. We should use this to show urgency that a decision needs to be made so efforts can begin making it happen if the goal is to help capture D7 folks looking to migrate.

This discussion feels like it needs some urgency, but as folks have pointed out, BD has working contribs for both of the D7 variants which means there is really no urgency after all...

I bring this up because I'm stuck trying to figure out: "Which one do I use that will cause me the least headache later?" At the very least, I think a formal decision needs to be made and clarified with urgency so those of us who want to build out a site for the future, don't have to worry about choosing a path that will have bad potential side-effects later.

Based on all of the various discussions, it appears that the actual debate is between the BD "Reference" module and the D7 "Entity Reference" module. I think we can mostly agree that the D7 "References" module is off the table as a contender. (Yes, I realize the D7 modules are ported to BD; I mention D7 to point out where they originated)

Between BD Reference and D7 Entity Reference, it appears that Entity Reference is stepping up to be the winning choice. Is this the general consensus we can all agree on or are there further discussions that need to take place in favor of BD Reference?

**Comment by BWPanda** [comment]
@philsward D7 ER looks to be the best choice for new sites, unless you're willing to help test & improve B Reference. I agree there's no urgency for a core reference-type module, so I suggest waiting for Reference to be finished, including upgrade paths from the D7 modules, and then have that as core's reference-type functionality.

One more argument for Reference in core is that work has already been started on it. Choosing ER instead will void all of the work done on Reference so far. Whereas the work done on ER will never be voided as it'll live on in contrib regardless (I don't see a Reference living on in contrib should ER be chosen instead).

**Comment by philsward** [comment]
Is Mike McCaffrey still around? It appears he was the main contributor in the early days. Very few commits have taken place recently.

From @hosef on Zulip: (I believe I connected the right person)

> As far as I was aware there were 4 objectives of the reference module:
- Have a UI that is easy to understand for non technical users
- Not use the namespace of an existing reference module
- Not depend on other contrib modules
- Cover all the use cases of existing contrib modules

> The code complexity/quality is not at issue here and it needs to stop coming up because it is holding up progress on discussions.

> I don't have a problem with the entity reference module and it's what I use in D7, but it we are going to include it, it will need to be modified so it doesn't depend on entity plus and so the UI makes sense to users with no knowledge of what entities are.

**Comment by philsward** [comment]
Rereading through the most recent comments, there is an overwhelming agreement to abandon the Reference module in favor of Entity Reference.

The decision by the community is very clear and doesn't appear to need any more debate.

Like @jenlampton mentioned, the next step is to create an issue that outlines the work necessary to make it happen.

Who is the most familiar with ER and could come up with a task list to follow in the new issue?

**Comment by stpaultim** [comment]
A little over 2 weeks ago, I posted an update suggesting that we try to push the Reference Module forward with the hope that it MIGHT get into 1.21.0 (or at least that we MIGHT make some substantial progress in that direction). I was pushing this option, because I THOUGHT we had consensus that this was the best solution AND because the consensus at the development meetings appeared to suggest that the Reference module was nearly ready to include in core (with some scaled down expectations). 

Since then a number of people have vocally expressed concerns that the Reference module (in their opinion) is NOT close to being ready and that the Entity Reference module is a better candidate anyway. The discussion in this issue (among those participating) moved dramatically in that direction, while all discussion in our weekly dev meetings suggests that the general policy has not changed and that Reference is still the preferred solution. See recording of meeting on Nov 11th (https://www.youtube.com/watch?v=tb767sTTgjA) - the first 20 minutes was spent on this issue. 

As far as I can tell, only one core maintainer is advocating for Entity Reference, while at least two core maintainers still (in my reading of the situation) seem to prefer the Reference module. 

I would love to see the core maintainers discuss this issue and come to some consensus. Until that happens, I would say that this issue has stalled and is unlikely to get much attention as long as there is no consensus on which direction we are moving.

**Comment by jenlampton** [comment]
We could escalate this to the PMC, and ask them to get a decision from core maintainers.

**Comment by BWPanda** [comment]
I have my opinion (expressed above), but whether or not the PMC agrees, it'd be good to get them to make a decision (based on core committer's suggestions and their own deliberations) so we can move forward either way.

**Comment by herbdool** [comment]
Here's a branch of Entity Reference that no longer requires entity_plus: https://github.com/backdrop-contrib/entityreference/tree/issue-21c. All the tests are passing. (Both ER and R would still benefit from a `view` and `buildContent` method in core to make the code better.

By the way, the high level discussion is not always helpful. The assumptions have been that ER is full of "legacy code" and R is fresh and new. But take a look at these two files, for example: https://github.com/backdrop-contrib/entityreference/blob/issue-21c/views/entityreference_plugin_display.inc and https://github.com/backdrop-contrib/reference/blob/1.x-1.x/views/reference_plugin_display.inc. The file in Reference is almost an exact copy from Entity Reference. Good. Copy and clean-up.

As @hosef said above: "If people now think that the entityreference module is stable enough to be the default reference solution, we could just rename the functions, update the UI, and call it a day."

I'm all for it.

**Comment by stpaultim** [comment]
> By the way, the high level discussion is not always helpful. The assumptions have been that ER is full of "legacy code" and R is fresh and new.

I believe that this is a vast over simplification of the arguments I've heard against Entity Reference. However, I am not really in a position to contribute to this discussion very much. I have a strong suspicion that a real time discussion of core committers could resolve this issue. It seems as if some folks who were initially resistant to Entity Reference as a solution have softened their positions. Although, I'm still not sure we have consensus. 

I believe pretty strongly that a real time discussion sometime in the next month, might position us to move this issue forward substantially during this release cycle. It is my sense that this issue is too complicated to resolve in the issue queue alone. 

I think it would be much better to get a consensus from core committers rather than take this to the PMC.

**Comment by stpaultim** [comment]
Interesting comment in the issue queue:
https://forum.backdropcms.org/comment/3743#comment-3743

**Comment by stpaultim** [comment]
Recently the discussion seems to have swung in the direction of Entity Reference. But, I'm still unclear where this is going.

**Comment by quicksketch** [comment]
I did a code review of Entity Reference today, and my opinion is that the module is well situated to be pulled into core with minimal changes. It no longer has any dependencies (like CTools, Entity API, or Entity Plus). It leverages a plugin system similar to Views and Layout modules, with classes that provide grouped collections of functionality.

At this point I think it would be worthwhile to start working with a PR that puts entityreference module in core. Creating the PR we should try to maintain the project git history, as we did with Views when it was merged into core. And then we'll see about the other changes needed to incorporate it, such as changing the default install profile, adding to `backdrop_merged_modules()`, and adding any update hooks needed.

We still have not "officially" decided on using entityreference, but we have agreed during weekly meeting that entityreference is the most used, most flexible, and most depended-upon solution we have. At this point, I think it would also be the easiest solution to integrate, simply because it already has all the features we need.

**Comment by herbdool** [comment]
Thanks @quicksketch for reviewing the module. I'll attempt a pr and see what people think.

**Comment by docwilmot** [comment]
What should be the approach for handling `taxonomy_term_reference` though? We cant remove it; contrib may be implementing Field API alter hooks etc. Options?
- Leave it in and have both `Term reference` and `Entity reference` available in Field API?
- Leave it in and somehow "hide" `Term reference` from the UI and somehow "redirect" API calls to `Entity reference`?
- Remove `Term reference` and upgrade path?
- Other?

Or maybe this isn't a  problem? I dont know Field API that well.

**Comment by quicksketch** [comment]
> What should be the approach for handling taxonomy_term_reference though?

I was thinking something along these lines:

1. Add Entity Reference to core.
2. Separate the Term Reference field from Taxonomy module.
3. Update install profiles to use Entity Reference.
4. Disable Term Reference module by default.
5. Provide _either_ a core conversion tool or contrib module to convert from term reference to entity reference.
6. In Backdrop 2.x: Remove Term Reference module. Move to contrib.

**Comment by quicksketch** [comment]
There's actually a whole issue about this question of replacing Term Reference at https://github.com/backdrop/backdrop-issues/issues/51.

**Comment by quicksketch** [comment]
I think the only big hold-up here is gracefully handling both the contrib and core module at the same time. Since we've done this before with other modules (like Link, Email, Redirect) we can repeat the same approach.

I pushed https://github.com/backdrop/backdrop/pull/4033/commits/0d907a6d6dfb8bd9dd735424801098ae4370c37f to the PR at https://github.com/backdrop/backdrop/pull/4033 that links to a draft change notice at https://docs.backdropcms.org/node/47321 (unpublished).

I tested downloading the contrib module, installing it, then switching to this branch. The update hook and status report both discovered the contrib version and linked to the change notice.

I did additional testing, tried out different widgets, reference methods, sorting, display modes... everything works great. There are some features I think are a little excessive (such as sorting on any arbitrary column from the base entity table), but if we maintain compatibility with the contrib version we need to continue supporting such options.

There is also test coverage and all tests are passing. Great job @herbdool!

This could use further review, but my impression is we are quite close with this approach.

**Comment by quicksketch** [comment]
I updated the summary with a more accurate representation of the current approach.

@herbdool Would you be willing to advocate for this issue so we can move it into the 1.23.0 milestone?

**Comment by herbdool** [comment]
@quicksketch thanks! I can advocate for this.

Are you thinking merging this is number 1 on your list and then number 2 through 4 on your list would be on the other issue #51?

**Comment by quicksketch** [comment]
> Are you thinking merging this is number 1 on your list and then number 2 through 4 on your list would be on the other issue https://github.com/backdrop/backdrop-issues/issues/51?

@herbdool Yes. Moving an entire module history into core like this is too unwieldy to try and wrap more into a single issue.

**Comment by herbdool** [comment]
Here's a follow-up to separate out the term reference field from the taxonomy module: https://github.com/backdrop/backdrop-issues/issues/5633

**Comment by indigoxela** [comment]
There's currently a failing test and it's not a random one. I verified by also running that test in the sandbox.

Entity Reference: Entity Reference Handlers (EntityReferenceHandlersTestCase)

So I'm setting the issue back to "needs work".

**Comment by indigoxela** [comment]
So I played with the PR sandbox - this all looks great! ... Until one tries to create a view of type "Entity reference" on a content type with an entity reference to a term ("category") and tries to filter the view by that category (term name) based on a relationship. View ui does odd things then (input field disappears). And the result stays empty.

Filtering by tid (numeric field value) works like a charm.

My description may be a bit confusing, but using views results for reference fields isn't that uncommon.

This may be handled in a follow-up, if people agree. The broken functional test should get fixed, though, before merging.

**Comment by argiepiano** [comment]
It'll be great to have this in core! Thanks for the work.

There are 3 PRs in the Entity Reference queue (2 of which I created in the last few days) which are in my view important fixes for issues related to third modules creating behavior and selection handlers for Entity Reference. The changes are minimal, but are important and I'm hoping they will make it to this PR before merging into core

**Comment by herbdool** [comment]
I've merged in the PRs @argiepiano mentions -- into the contrib module. Now I have to remember how I set up the git repo to merge it into core and update the history here too.

**Comment by quicksketch** [comment]
@herbdool If it would make things easier, I'm open to merging the current PR without the 3 additional commits. Then we can apply each one of the commits to the core repository directly. Would that work as an approach?

**Comment by herbdool** [comment]
@quicksketch yes, I think that sounds easier. Let's do it! Can make a follow-up for those commits.

**Comment by quicksketch** [comment]
I'll request any concerns from the leadership team (PMC + core committers) via Zulip and here. Any last concerns please speak now.

**Comment by indigoxela** [comment]
Plus one for merging these 4,508 lines of code and in follow-ups iterate over all quirks, bugs, oddities found while testing. Doing it step by step seems easier for all people involved.

But I do have one request: EntityReferenceHandlersTestCase (Tests for the base handlers provided by Entity Reference.) is still failing. IMP it's better to fix the test before merging. (In other words, let's not make a precedent for merging with failing tests :wink:)

**Comment by herbdool** [comment]
When I run the tests locally, they all pass. So this is odd and hard to debug. I can try restart the tests here just to see.

**Comment by indigoxela** [comment]
> When I run the tests locally, they all pass.

Very strange. Is your local install "vanilla", or are modules installed that aren't there in the sandbox (where the test also fails)?

The first thing to try may be a rebase (just in case...).

**Comment by argiepiano** [comment]
I've been able to reproduce the test failure in my local, and will take a closer look today

**Comment by herbdool** [comment]
It's the new card content. I've updated the test to remove all the default nodes (before it was removing the original 2 nodes, now 5).

**Comment by argiepiano** [comment]
That's great. Isn't it easier to use the testing profile? That skips creating the default nodes

**Comment by herbdool** [comment]
Usually yes. In this case I didn't want to change the old tests too much.

Locally I was able to recreate the failures and failures are gone when I delete the nodes. But tests on github are still failing.

**Comment by herbdool** [comment]
All tests passing except one unrelated one in BootstrapPageCacheTestCase.

**Comment by stpaultim** [comment]
It looks to me like we have a version right now with all tests passing! Whoopeee....

**Comment by quicksketch** [comment]
Merged https://github.com/backdrop/backdrop/pull/4033 in 1.x for 1.23.0. Wow! Thanks @herbdool for putting together this final PR, advocating for Entity Reference as the solution, and your continuous work on Entity Reference in contrib leading up to this!

[73c65a0](https://github.com/backdrop/backdrop/commit/73c65a0865f9e3dea79d5dba0ec385d3374b96a5) by @herbdool, @argiepiano, @klonos, @indigoxela, @jenlampton, @docwilmot, @stpaultim, @philsward, @quicksketch, @mikemccaffrey, @BWPanda, @laryn, @hosef, @serundeputy, @olafgrabienski, @MrHaroldA, @keiserjb, and @jromine.

I hope I got everyone that contributed to this issue, even if was in the decision-making process or experimentation with our other options (thanks @mikemccaffrey and @jromine in particular).

**Comment by herbdool** [comment]
I also created a PR for https://github.com/backdrop/backdrop-issues/issues/2300, fixing the `File::uri()`. Now that entityreference is in core, I changed the other issue to a bug.
